src/main/java/calculator/Tokens.java

```diff
@@ -0,0 +1,75 @@
+package calculator;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.stream.Collectors;
+
+import static calculator.Calculator.NUMBER_REGEX;
+import static calculator.Calculator.SIGN_REGEX;
+
+public class Tokens implements Iterable<Token> {
+    private final List<Token> tokens;
+
+    public Tokens(final String[] tokens) {
+        validateExpression(tokens);
+        this.tokens = Arrays.stream(tokens)
+            .map(Token::new)
+            .collect(Collectors.toUnmodifiableList());
+    }
+
+    private static void validateExpression(final String[] tokens) {
+        validateTokenCountIsOdd(tokens);
+        validateTokensByPosition(tokens);
+    }
+
+    private static void validateTokenCountIsOdd(final String[] tokens) {
+        if (tokens.length % 2 == 0) {
+            throw new IllegalArgumentException("수식에 오류가 있습니다.");
+        }
+    }
+
+    private static void validateTokensByPosition(final String[] tokens) {
+        for (int i = 0; i < tokens.length; i++) {
+            validateToken(tokens[i], i);
+        }
+    }
+
+    private static void validateToken(final String token, final int index) {
+        if (isEven(index)) {
+            validateNumberToken(token);
+        }
+
+        if (!isEven(index)) {
+            validateSignToken(token);
+        }
```

Comment:
```suggestion
        if (isEven(index)) {
            validateNumberToken(token);
            return;
        }
        validateSignToken(token);
```
if문을 두지 않고 바로 return 처리하면 더 깔끔할 것 같네요 😄